MRG: Do regularization in PCA space

larsoner opened the pull request:
We currently do covariance regularization on the full rank data. Thus rank-deficient covariance matrices become full rank. This PR changes it so that regularization is done in the lower-dimensional space, then projected back out, which should preserve the original rank.

This probably has some small impact on full-rank MEG data (e.g., 306 channels with a handful of projs) but the impact is magnified by Maxwell filtering due to the larger rank reduction.

- [x] ~~Something is fishy with~~ the `loglik` values are pretty uniformly `-39` in the previous tests across all methods, because they all provide the same quality estimates when using low-rank computation. This score was the best score before, so it's reasonable they all achieve it now (assuming the differences before were driven by how well the low-rank-ness was captured by each method).
- [x] Need to check examples
- [x] Try on SSS'ed data
- [x] Add check that `mne.cov.regularize(..., proj=True)` gives same result as `diagonal_fixed` now
- [x] Fix `mne.cov.regularize` to respect rank with new `rank` argument
- [x] Fix defaults for `diagonal_fixed` method to match those of `mne.cov.regularize`
- [x] ~~Fix some missing FIFF constants, and fix check for missing ones~~ #5691
- [x] ~~Fix `job` output in info for `maxwell_filtering` to reflect what we actually do~~ #5691
- [x] Make CIs happy
- [x] Check diagonal_fixed result -- amplitude decreases in plot_white example?
- [x] Merge #5691 and rebase this
- [x] Update API to reflect beamformer changes

Updated examples (930bee9):

- [PR](https://10142-1301584-gh.circle-artifacts.com/0/html/auto_examples/inverse/plot_covariance_whitening_dspm.html) - [master](http://mne-tools.github.io/dev/auto_examples/inverse/plot_covariance_whitening_dspm.html)
- [PR](https://10142-1301584-gh.circle-artifacts.com/0/html/auto_examples/inverse/plot_lcmv_beamformer.html) - [master](http://mne-tools.github.io/dev/auto_examples/inverse/plot_lcmv_beamformer.html)
- [PR](https://10142-1301584-gh.circle-artifacts.com/0/html/auto_examples/visualization/plot_evoked_whitening.html) - [master](http://mne-tools.github.io/dev/auto_examples/visualization/plot_evoked_whitening.html)
- [PR](https://10142-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_brainstorm_phantom_elekta.html) - [master](http://mne-tools.github.io/dev/auto_tutorials/plot_brainstorm_phantom_elekta.html)
- [PR](https://10142-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_compute_covariance.html) - [master](http://mne-tools.github.io/dev/auto_tutorials/plot_compute_covariance.html)
- [PR](https://10142-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_whitened.html) - [master](http://mne-tools.github.io/dev/auto_tutorials/plot_whitened.html)

It looks like most did not change. I did change `bst_phantom_elekta` to use `method='shrunk'`, which no longer increases the cov rank to 306 (hooray!).

wmvanvliet commented:
Can you explain what are the benefits exactly? Sounds like you've thought this through and came to some insight.

wmvanvliet commented:
Yes, it makes sense, do you see a big difference in the inverse results?

larsoner commented:
Okay I think this is ready for review/merge. We should probably wait for 0.18 just to be safe / have time to test it, though, so I'm going to leave set as WIP until we release.

larsoner commented:
On second thought, I think the tests are pretty convincing, so I'd be happy putting this in 0.17. Let's see if @agramfort @dengemann you are convinced, too :)

larsoner commented:
Updated examples:

- https://10065-1301584-gh.circle-artifacts.com/0/html/auto_examples/inverse/plot_covariance_whitening_dspm.html
- https://10065-1301584-gh.circle-artifacts.com/0/html/auto_examples/inverse/plot_lcmv_beamformer.html
- https://10065-1301584-gh.circle-artifacts.com/0/html/auto_examples/visualization/plot_evoked_whitening.html
- https://10065-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_brainstorm_phantom_elekta.html
- https://10065-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_compute_covariance.html
- https://10065-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_whitened.html

It looks like most did not change. I did change `bst_phantom_elekta` to use `method='shrunk'`, which no longer increases the cov rank to 306 (hooray!).

larsoner commented:
Okay I also fixed the `diagonal_fixed` estimator defaults in `compute_covariance` to match those of `mne.cov.regularize` (I think it was a bug that they did not -- `eeg=0.` by default for example). I also fixed the `FIFF` constant checks (and added some constants) because I realized we were missing some for Maxwell filtering.

larsoner commented:
Talking to @dengemann to be safe we should probably wait for 0.18 to merge this. So I'll set to WIP for now. (Also apparently Travis is unhappy so I have some work to do anyway.) Also I'm not 100% sure the `diagonal_fixed` result is correct. Need to check.

larsoner commented:
Fixed the "problem" with the `plot_evoked_whitening.py` example. The differences were caused in the changes in defaults, which seems okay. URLs above will updated soon to reflect commit a30e48b.

larsoner commented:
It probably makes sense to get #5146 in after this, as it makes clear exactly what is happening to your data rank at each step (SSP, cov reg, etc.)

dengemann commented on mne/cov.py:
```diff
@@ -545,14 +554,45 @@ def _check_method_params(method, method_params, keep_sample_mean=True,
+        if was_auto:
+            method.pop(method.index('factor_analysis'))
+        for m in method:
+            if m in ('pca', 'factor_analysis'):
```
m -> mm

dengemann commented on mne/cov.py:
```diff
@@ -545,14 +554,45 @@ def _check_method_params(method, method_params, keep_sample_mean=True,
+
+
+def _check_rank(rank, method, was_auto=False):
+    if isinstance(rank, string_types):
```
some simple doc string ?

dengemann replied:
@larsoner Can you verbally describe what is the goal here. The moment where I see code from the `operator' module in use I start getting worried ...

larsoner replied:
`operator.index` is just the preferred way to check if something is an int

larsoner replied:
... and the `''` is just a way to deprecate `rank='full'` (current `master` behavior) to `rank=None` (auto-detect and preserve)

dengemann commented on mne/cov.py:
```diff
@@ -876,67 +937,83 @@ def _check_scalings_user(scalings):
             est = EmpiricalCovariance(**mp)
             est.fit(data_)
-            estimator_cov_info.append((est, est.covariance_, _info))
+            estimator_cov_info.append([est, est.covariance_, None])
```
why do we convert those to lists?

larsoner replied:
Might not be necessary (can't remember if I needed it at one point), but I also don't mind it because visually it helps separate the append method parens from what is being appended

larsoner replied:
Ahh yes it is necessary:
```
        # undo scaling
        for c in estimator_cov_info:
            if eigvec is not None:
                # project back if necessary
>               c[1] = np.dot(eigvec.T, np.dot(c[1], eigvec))
```

dengemann replied:
Ok cool. Just wanted to make sure we have some reasons.

dengemann commented:
@larsoner this looks terrific. Do we have some SSS data example here that shows some more drastic low rank computation? I feel we'd need some example that covers the complaints we received from maxfilter users.

dengemann commented:
@wmvanvliet FYI this is the continuation of my #4637 that we abandoned last year as we had a few doubts that we could not remove just before the release. @larsoner kept his promise to look into it if I would not insist merging it. Awesome we're getting there! I will run the current code tomorrow with a few pathological SSS recordings I collected over the last 2 years to see where this gets us.

larsoner commented:
> Do we have some SSS data example here that shows some more drastic low rank computation? I feel we'd need some example that covers the complaints we received from maxfilter users.

Yes I modified this example already:

https://10075-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_brainstorm_phantom_elekta.html

dengemann commented:
@larsoner thank you! I'll checkout diverse SSSd data with it tomorrow. Let's see where we get, perhaps we can merge it before the end of the week.

dengemann commented on mne/cov.py:
```diff
@@ -876,24 +938,41 @@ def _check_scalings_user(scalings):
     """Compute covariance auto mode."""
     # rescale to improve numerical stability
     _apply_scaling_array(data.T, picks_list=picks_list, scalings=scalings)
+    if rank != 'full':
```
Should be a better test. If len(data) == 306 and rank == 306 this block should not be entered.

larsoner replied:
A couple of things:

1. The check for this would actually be a bit annoying, as you need to then triage based on the various incantations of `rank` input, instead of just letting `_smart_eigh` be the one place this is handled / estimated.
2. In principle it should not matter. As long as we do the rotation properly, it should gracefully fall back to identical behavior.

I have now added a test that point (2) holds. It turns out that it did not before (whoops) but it does now!

dengemann commented on doc/whats_new.rst:
```diff
@@ -25,6 +25,8 @@ Changelog
 - Add :func:`mne.io.read_raw_fieldtrip`, :func:`mne.read_epochs_fieldtrip` and :func:`mne.read_evoked_fieldtrip` to import FieldTrip data. By `Thomas Hartmann`_ and `Dirk Gütlin`_.
 
+- Add ``rank`` parameter to :func:`mne.compute_covariance`, :func:`mne.cov.regularize` and related functions to preserve data rank during regularization by `Eric Larson`_ and `Denis Engemann`_
+
```
Can you also add that we now support faster low-rank (PCA-space) computation?

larsoner replied:
Pushed a commit, if it's not what you want feel free to give me a more complete suggestion or push a change

I avoided the word "PCA" because it's more about dealing with rank deficiency rather than projecting onto (orthogonal) directions of highest variance

dengemann replied:
One could still highlight that the option can make things significantly faster for SSSd data.

larsoner replied:
"and speed up computation" is already in there

dengemann commented:
After playing with several datasets my impression is that this PR does not make things worse than they are. Instead for certain SSS'd the default looks now correct, hence things are improved. As noticed before, the speedup is practically meaningful. I'd suggest to merge this and then work on the outstanding issues in #4676 which are not magically addressed by this PR -- just for expectation management. I'll approve and suggest @agramfort to take his tour.

agramfort commented:
pushed a cosmit.

now with the inverse method the inverse problem are automagically solved in low rank space? or we still have square whitener even if rank is low?

larsoner commented:
This does not change the behavior of inverse operators / LCMV. So it's still stored square IIRC though at some point gets reduced internally (at least for minimum norm)

larsoner commented:
Your cosmit broke a test

agramfort commented:
to get your forgiveness (if I may ask) let me merge this one...
